README.md
# scss-to-css
> It can be used for simple compilation/compression of scss files without installing various front-end engineering tools (webpack, etc.).

[![Version](https://vsmarketplacebadge.apphb.com/version-short/yutent.scss-to-css.svg)](https://marketplace.visualstudio.com/items?itemName=yutent.scss-to-css)
[![Rating](https://vsmarketplacebadge.apphb.com/rating-short/yutent.scss-to-css.svg)](https://marketplace.visualstudio.com/items?itemName=yutent.scss-to-css)
[![Installs](https://vsmarketplacebadge.apphb.com/installs/yutent.scss-to-css.svg)](https://marketplace.visualstudio.com/items?itemName=yutent.scss-to-css)
[![Build Status](https://travis-ci.org/yutent/scss-to-css.svg?branch=master)](https://travis-ci.org/yutent/scss-to-css)


[README_中文](./README_ZH.md)


Live demo:
![demo](./demo.gif)

## Why Scss-to-css
> For small projects, various engineering tools such as webpack are too heavy and cumbersome to configure. And a lot of modules have to be installed.
> Sometimes, we just want to simply use the convenience brought by scss. So for this purpose, I wrote a vsc plug-in, which can be automatically compiled into a css file (stored in the current directory by default) when the scss file is saved, and the browser prefix is ​​automatically completed.


## Configuration
> Some configuration can be set, which it works better for you。
>> - `compileOnSave`: Auto compile on document saved, default `true`
>> - `autoPrefixer`: Will autoprefixer. It can be run faster when turn off. default `true`
>> - `output`: Output style. default `compressed`。 
>> - `exclude`: The RegExp of path what you can to ignore(the `var.scss file` will never be compiled)。


## Issues
> This extension work well on all os. If any problem please let me known [issue](https://github.com/yutent/scss-to-css/issues)


## Dependencies
> We use `node-sass`(Deprecated in 3.1.0) instead of `libsass`.
>> v3.1.0 uses `sass` instead. No need to install node-sass globally

 - `node-sass`, You need to install this module manually. Maybe `root` is required on linux.


## Installation
> Search `scss-to-css` and install in the marketplace.


## .browserslistrc DEMO (Deprecated in 2.x)
> Just for demo, you can change it by youself. If not exists, the default value will be `last 2 version`.

```
ie > 9
iOS > 8
Android >= 4.4
ff > 38
Chrome > 38
```

## .scssrc DEMO (in 2.x or above)
> instead of using `.browserslistrc`, we recommend to use `.scssrc`.

```json
{
  "browsers": ["ie > 9", "iOS > 8", "Android >= 4.4", "ff > 38", "Chrome > 38"],
  "outdir": "dist" // relative path of this '.scssrc' file.
}

// You can also set other config in this file.
// eg. compileOnSave,autoPrefixer,output,exclude
```

CHANGELOG.md
# 更新日志(Changed Logs)

## [3.0.2] 2021-06-30
- 改用`sass`, 无需全局安装`node-sass`了

## [3.0.2] 2020-08-03
- 代码优化

## [3.0.0] 2019-08-05
- 删除libsass, 改用`node-sass`(需要手动全局安装)


## [2.1.0] 2019-06-10
- 更新libsass到最新版
- 增加异常信息输出


## [2.0.0] 2019-02-14
- 大版本改动,移除 `.browserslistrc`的支持, 改用`.scssrc`;
- 新增`outdir`选项支持, 允许将编译后的css文件统一存放.


## [1.2.0] 2019-01-05
- 配置优化


## [1.1.1] 2018-12-27
- 更换sass库,兼容win10 [#1](https://github.com/yutent/scss-to-css/issues/1)


## [1.1.0] 2018-12-21
- 增加右键菜单编译(此命令不受配置中的过滤条件限制,只判断是否scss文件)


## [1.0.4] 2018-12-17
- 日常维护


## [1.0.3] 2018-12-03
- 修复autoprefixer的配置读取


## [1.0.0] 2018-12-03
- 正式上架插件商店。

test/index.js
const testRunner = require('vscode/lib/testrunner')

testRunner.configure({
  ui: 'tdd', // the TDD UI is being used in extension.test.ts (suite, test, etc.)
  useColors: true, // colored output from test results
  timeout: 10000
})

module.exports = testRunner


index.js
/**
 *
 * @author yutent<yutent.io@gmail.com>
 * @date 2018/11/01 09:37:55
 */

const vsc = require('vscode')
const path = require('path')
const scss = require('sass')

const fs = require('iofs')
const postcss = require('postcss')
const autoprefixer = require('autoprefixer')
let prefixer

const std = vsc.window.createOutputChannel('scss-to-css')
std.out = function(msg) {
  std.appendLine(msg)
}

// 编译scss
function render(style, file) {
  try {
    let data = fs.cat(file).toString()
    return (
      scss.renderSync({
        data,
        outputStyle: style
      }).css + ''
    ).trim()
  } catch (err) {
    std.out(err)
    // console.error(err)
  }
}

let options = {
  compileOnSave: true,
  autoPrefixer: true,
  output: 'compressed',
  exclude: ''
}

const compileScss = (style, entry, output) => {
  if (options.outdir) {
    let tmp = output.replace(options.workspace, '.')
    output = path.join(options.outdir, tmp)
  }

  let css = render(style, entry)

  if (options.autoPrefixer) {
    return prefixer.process(css, { from: '', to: '' }).then(result => {
      return { css: result.css, output }
    })
  } else {
    return Promise.resolve({ css, output })
  }
}

const Compiler = {
  compile(doc) {
    let origin = doc.fileName || ''
    let target = origin.replace(/\.scss$/, '')
    let task = []

    // 说明不是scss文件
    if (origin === target) {
      return
    }

    task = options.output.map(style => {
      let ext = '.css'

      switch (style) {
        case 'compressed':
          ext = '.min' + ext
          break
        default:
          ext = style.slice(0, 1) + ext
      }

      return { style, output: target + ext }
    })

    // 编译单一类型, 则去掉文件名微调
    if (task.length === 1) {
      task[0].output = target + '.css'
    }

    task = task.map(item => {
      return compileScss(item.style, origin, item.output)
    })

    Promise.all(task)
      .then(list => {
        list.forEach(it => {
          fs.echo(it.css, it.output)
        })
      })
      .catch(err => {
        vsc.window.showInformationMessage(err)
      })
  },

  /**
   * 条件过滤
   * 用于保存时编译的动作, 右键编译时, 不过滤这2项
   */
  filter(doc) {
    // 未开启保存时编译
    if (!options.compileOnSave) {
      return
    }

    let origin = doc.fileName || ''

    // var.scss文件默认不编译
    if (/\/var\.scss$/.test(origin)) {
      return
    }

    // 过滤不编译的文件
    if (options.exclude) {
      if (options.exclude.test(origin)) {
        return
      }
    }

    this.compile(doc)
  }
}

function __init__() {
  let conf = vsc.workspace.getConfiguration('Scss2css')
  let folders = vsc.workspace.workspaceFolders
  let wsDir = ''
  let configFile = ''

  Object.assign(options, conf)
  conf = null

  options.output = options.output.split('|').map(it => it.trim())

  if (folders && folders.length) {
    wsDir = folders[0].uri.path
  }

  if (wsDir) {
    configFile = path.join(wsDir, '.scssrc')
  } else {
    let editor = vsc.window.activeTextEditor
    if (editor) {
      wsDir = path.dirname(editor.document.fileName)
      configFile = path.join(wsDir, '.scssrc')
    }
  }

  options.workspace = wsDir

  // 有配置文件时, 优先使用配置文件的配置
  if (fs.exists(configFile)) {
    let tmp = JSON.parse(fs.cat(configFile).toString())

    Object.assign(options, tmp)
    tmp = null

    if (options.outdir) {
      options.outdir = path.join(options.workspace, options.outdir)
    }
  }

  if (options.exclude) {
    options.exclude = new RegExp(options.exclude, 'i')
  }

  prefixer = postcss().use(
    autoprefixer({
      browsers: options.browsers
    })
  )
}

function deactivate() {}

exports.activate = function(ctx) {
  __init__()

  vsc.workspace.onDidChangeConfiguration(__init__)

  vsc.workspace.onDidSaveTextDocument(doc => {
    std.clear()
    Compiler.filter(doc)
  })

  let cmd = vsc.commands.registerCommand('Scss2css.compile', _ => {
    let editor = vsc.window.activeTextEditor

    if (editor) {
      Compiler.compile(editor.document)
    }
  })
  ctx.subscriptions.push(cmd)
}
exports.deactivate = deactivate


scss-compiler.py
import sublime
import sublime_plugin
import subprocess
import re
import os


def get_setting(key, default=None):

    settings = sublime.load_settings('ScssCompiler.sublime-settings')
    os_specific_settings = {}

    return os_specific_settings.get(key, settings.get(key, default))


def compile_scss(scss_file): 

    shoud_compress = get_setting('compress_on_compile')

    compile_style = get_setting('compile_style')

    sourcemap = get_setting('sourcemap')


    args = ['scss', '-C', '-t']
    
    if shoud_compress:
        args.append('compressed')
    else:
        args.append(compile_style)

    if sourcemap:
        args.append('--sourcemap=auto')
    else:
        args.append('--sourcemap=none')

    args.append(scss_file)
    
    cssfile = scss_file[0:-4] + 'css'
    
    args.append(cssfile)

    pipe = subprocess.Popen(args, stderr=subprocess.PIPE)
    (out, err) = pipe.communicate()
    return err.decode('utf-8')
    


class ScssCompiler(sublime_plugin.EventListener):
    def on_post_save(self, view):

        shoud_compile = get_setting('compile_on_save')
        scss_file = view.file_name()

        if not shoud_compile: 
            return

        if not re.search(r'\.scss$', scss_file): 
            return 

        if os.path.basename(scss_file) == 'var.scss': 
            return 
        
        err_msg = compile_scss(scss_file)

        if err_msg: 
            view.show_popup(err_msg)


class __CompilerCommand(): 
    def get_path(self, paths): 
        if paths: 
            return paths[0]
        elif self.window.active_view() and self.window.active_view().file_name(): 
            return self.window.active_view().file_name()
        elif self.window.folders(): 
            return self.window.folders()
        else:
            sublime.error_message('No scss file to Compile...')
            return False


class CompileScssCommand(sublime_plugin.WindowCommand, __CompilerCommand): 

    def run(self, paths=[]): 

        path = self.get_path(paths)
        if not path: 
            return 

        print(path)

        if not re.search(r'\.scss$', path): 
            sublime.error_message('This is not a scss file ...')
            return 

        compile_scss(path)
        

entry.js
/**
 *
 * @author yutent<yutent.io@gmail.com>
 * @date 2018/11/01 09:37:55
 */

const vsc = require('vscode')
const path = require('path')
const scss = require('sass')

const fs = require('iofs')
const postcss = require('postcss')
const autoprefixer = require('autoprefixer')
let prefixer

const std = vsc.window.createOutputChannel('scss-to-css')
std.out = function(msg) {
  std.appendLine(msg)
}

// 编译scss
function render(style, file) {
  try {
    let data = fs.cat(file).toString()
    return (
      scss.renderSync({
        data,
        outputStyle: style
      }).css + ''
    ).trim()
  } catch (err) {
    std.out(err)
    // console.error(err)
  }
}

let options = {
  compileOnSave: true,
  autoPrefixer: true,
  output: 'compressed',
  exclude: ''
}

const compileScss = (style, entry, output) => {
  if (options.outdir) {
    let tmp = output.replace(options.workspace, '.')
    output = path.join(options.outdir, tmp)
  }

  let css = render(style, entry)

  if (options.autoPrefixer) {
    return prefixer.process(css, { from: '', to: '' }).then(result => {
      return { css: result.css, output }
    })
  } else {
    return Promise.resolve({ css, output })
  }
}

const Compiler = {
  compile(doc) {
    let origin = doc.fileName || ''
    let target = origin.replace(/\.scss$/, '')
    let task = []

    console.log(origin, target)

    // 说明不是scss文件
    if (origin === target) {
      return
    }

    task = options.output.map(style => {
      let ext = '.css'

      switch (style) {
        case 'compressed':
          ext = '.min' + ext
          break
        default:
          ext = style.slice(0, 1) + ext
      }

      return { style, output: target + ext }
    })

    // 编译单一类型, 则去掉文件名微调
    if (task.length === 1) {
      task[0].output = target + '.css'
    }

    task = task.map(item => {
      return compileScss(item.style, origin, item.output)
    })

    Promise.all(task)
      .then(list => {
        console.log('>'.repeat(50), '\n', list)
        list.forEach(it => {
          fs.echo(it.css, it.output)
        })
      })
      .catch(err => {
        vsc.window.showInformationMessage(err)
      })
  },

  /**
   * 条件过滤
   * 用于保存时编译的动作, 右键编译时, 不过滤这2项
   */
  filter(doc) {
    // 未开启保存时编译
    if (!options.compileOnSave) {
      return
    }

    let origin = doc.fileName || ''

    // var.scss文件默认不编译
    if (/\/var\.scss$/.test(origin)) {
      return
    }

    // 过滤不编译的文件
    if (options.exclude) {
      if (options.exclude.test(origin)) {
        return
      }
    }

    this.compile(doc)
  }
}

function __init__() {
  let conf = vsc.workspace.getConfiguration('Scss2css')
  let folders = vsc.workspace.workspaceFolders
  let wsDir = ''
  let configFile = ''

  Object.assign(options, conf)
  conf = null

  options.output = options.output.split('|').map(it => it.trim())

  if (folders && folders.length) {
    wsDir = folders[0].uri.path
  }

  if (wsDir) {
    configFile = path.join(wsDir, '.scssrc')
  } else {
    let editor = vsc.window.activeTextEditor
    if (editor) {
      wsDir = path.dirname(editor.document.fileName)
      configFile = path.join(wsDir, '.scssrc')
    }
  }

  options.workspace = wsDir

  // 有配置文件时, 优先使用配置文件的配置
  if (fs.exists(configFile)) {
    let tmp = JSON.parse(fs.cat(configFile).toString())

    Object.assign(options, tmp)
    tmp = null

    if (options.outdir) {
      options.outdir = path.join(options.workspace, options.outdir)
    }
  }

  if (options.exclude) {
    options.exclude = new RegExp(options.exclude, 'i')
  }

  prefixer = postcss().use(
    autoprefixer({
      browsers: options.browsers
    })
  )
}

function deactivate() {}

exports.activate = function(ctx) {
  __init__()

  vsc.workspace.onDidChangeConfiguration(__init__)

  vsc.workspace.onDidSaveTextDocument(doc => {
    std.clear()
    Compiler.filter(doc)
  })

  let cmd = vsc.commands.registerCommand('Scss2css.compile', _ => {
    let editor = vsc.window.activeTextEditor

    if (editor) {
      Compiler.compile(editor.document)
    }
  })
  ctx.subscriptions.push(cmd)
}
exports.deactivate = deactivate
